datajoint/expression.py

```diff
@@ -931,3 +957,20 @@ def aggr(self, group, **named_attributes):
         )
 
     aggregate = aggr  # alias for aggr
+
+
+def _flatten_attribute_list(primary_key, attrs):
+    """
+    :param primary_key: list of attributes in primary key
+    :param attrs: list of attribute names, which may include "KEY", "KEY DESC" or "KEY ASC"
+    :return: generator of attributes where "KEY" is replaced with its component attributes
+    """
+    for a in attrs:
```

Comment:
could you test this using attributes that are SQL reserved words, e.g. "select", "key", "index"? We may need to escape them.

Reply:
They do seem to cause an issue:
```
datajoint.errors.QuerySyntaxError: ("You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'select, key, index LIMIT 1) as `$0`' at line 1", 'SELECT count(DISTINCT `id_l`) FROM (SELECT `id_l`,`cond_in_l` FROM `djtest_relational`.`#l` ORDER BY select, key, index LIMIT 1) as `$0`')
```

Reply:
SQL has a ton of reserved words. We typically do not prevent the user from using these names in attribute names. This means that we should escape these attribute names. In MySQL this is done with backticks around the attribute names.

Reply:
In MySQL, attributes can be escaped using backquotes.

Reply:
Neither this:
```
@schema
class Keywords(dj.Lookup):
    definition = """  # tests sql keyword escaping
    select  : int
    key  : int
    ---
    index: int
    """
```
Nor this:
```
@schema
class Keywords(dj.Lookup):
    definition = """  # tests sql keyword escaping
    `select`  : int
    ---
    `index`: int
    """
```
Is working for me.

Reply:
What's the error message?  The first declaration should work as datajoint encloses the attribute names in backquotes. If that does not work, we should fix it.

Reply:
```
ERROR: Failure: AttributeError ('NoneType' object has no attribute 'groupdict')
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/nose/failure.py", line 39, in runTest
    raise self.exc_val.with_traceback(self.tb)
  File "/usr/local/lib/python3.7/site-packages/nose/loader.py", line 418, in loadTestsFromName
    addr.filename, addr.module)
  File "/usr/local/lib/python3.7/site-packages/nose/importer.py", line 47, in importFromPath
    return self.importFromDir(dir_path, fqname)
  File "/usr/local/lib/python3.7/site-packages/nose/importer.py", line 94, in importFromDir
    mod = load_module(part_fqname, fh, filename, desc)
  File "/usr/local/lib/python3.7/imp.py", line 234, in load_module
    return load_source(name, filename, file)
  File "/usr/local/lib/python3.7/imp.py", line 171, in load_source
    module = _load(spec)
  File "<frozen importlib._bootstrap>", line 696, in _load
  File "<frozen importlib._bootstrap>", line 677, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 728, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/workspaces/datajoint-python/tests_old/test_relational_operand.py", line 19, in <module>
    from .schema_simple import (
  File "/workspaces/datajoint-python/tests_old/schema_simple.py", line 18, in <module>
    class Keywords(dj.Lookup):
  File "/workspaces/datajoint-python/datajoint/schemas.py", line 177, in __call__
    self._decorate_master(cls, context)
  File "/workspaces/datajoint-python/datajoint/schemas.py", line 189, in _decorate_master
    cls, context=dict(context, self=cls, **{cls.__name__: cls})
  File "/workspaces/datajoint-python/datajoint/schemas.py", line 226, in _decorate_table
    instance.declare(context)
  File "/workspaces/datajoint-python/datajoint/table.py", line 96, in declare
    sql, external_stores = declare(self.full_table_name, self.definition, context)
  File "/workspaces/datajoint-python/datajoint/declare.py", line 310, in declare
    ) = prepare_declare(definition, context)
  File "/workspaces/datajoint-python/datajoint/declare.py", line 265, in prepare_declare
    compile_index(line, index_sql)
  File "/workspaces/datajoint-python/datajoint/declare.py", line 447, in compile_index
    r"(?P<unique>unique\s+)?index\s*\(\s*(?P<args>.*)\)", line, re.I
AttributeError: 'NoneType' object has no attribute 'groupdict'
```

Reply:
that's a regular expression mismatch, which we should replace with a more meaningful error by the way. Can you rethrow it as `DataJointError("table definition syntax error in line ....")`?